core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1300,18 +1301,29 @@ class UnifiedLog(@volatile var logStartOffset: Long,
       } else if (targetTimestamp == ListOffsetsRequest.EARLIEST_LOCAL_TIMESTAMP) {
         val curLocalLogStartOffset = localLogStartOffset()
 
-        val earliestLocalLogEpochEntry = leaderEpochCache.asJava.flatMap(cache => {
-          val epoch = cache.epochForOffset(curLocalLogStartOffset)
-          if (epoch.isPresent) cache.epochEntry(epoch.getAsInt) else Optional.empty[EpochEntry]()
-        })
-
-        val epochOpt = if (earliestLocalLogEpochEntry.isPresent && earliestLocalLogEpochEntry.get().startOffset <= curLocalLogStartOffset)
-          Optional.of[Integer](earliestLocalLogEpochEntry.get().epoch)
-        else Optional.empty[Integer]()
+        var epochResult: Optional[Integer] = Optional.empty()
```

Comment:
Can we use `val` instead of `var` here?

```
        val epochResult: Optional[Integer] =
          if (leaderEpochCache.isDefined) {
            val epochOpt = leaderEpochCache.get.epochForOffset(curLocalLogStartOffset)
            if (epochOpt.isPresent) Optional.of(epochOpt.getAsInt) else Optional.of(RecordBatch.NO_PARTITION_LEADER_EPOCH)
          } else Optional.of(RecordBatch.NO_PARTITION_LEADER_EPOCH)

```

Reply:
I believe the condition for both this and the below need to be slightly altered to accommodate for https://github.com/apache/kafka/pull/15213#discussion_r1481540437. I have since made the change so let me know whether it is okay with you 😊